File: C++/tests/TestMethod.cpp
#include <Method.h>
#include <Entity.h>
#include <EntityCore.h>

using namespace std;

EntityCore Core;

Component<float> LEFT 		= Core.newComponent("left",0.0f);
Component<float> RIGHT 		= Core.newComponent("right",0.0f);
EntityType* EXTENT 			= Core.newEntityType({LEFT.id, RIGHT.id}, {}, View::NONE);


void shrinkDefault(Entity &e, float amount) {
	float gap = (e.get(RIGHT) - e.get(LEFT)) * 0.5f;
	float center = gap + e.get(LEFT);
	e.set(LEFT, center - gap * amount);
	e.set(RIGHT, center + gap * amount);
}

Method<void(float)> SHRINK = Core.newMethod<void,float>("shrink", {LEFT.id, RIGHT.id}, shrinkDefault);

int main()
{
	EXTENT->addMethod(SHRINK);

	Entity e(EXTENT);

	e.set(LEFT, 2.0f);
	e.set(RIGHT, 6.0f);

	e.execute(SHRINK, 0.5f);

	cout << e(LEFT) << " -> " << e(RIGHT) << endl;

	return 0;
} 

File: Java/test/org/magnos/entity/TestMethod.java
package org.magnos.entity;

import static org.junit.Assert.*;

import org.junit.Test;
import org.magnos.entity.helper.Scalar;

public class TestMethod
{

	public static final EntityCore core = new EntityCore();

	public static final Component<Scalar> LEFT = core.newComponent( "left", new Scalar() );
	public static final Component<Scalar> RIGHT = core.newComponent( "right", new Scalar() );
	
	public static final Method<Void> SHRINK = core.newMethod( "shrink", new BitSet(LEFT, RIGHT), MethodReflection.staticMethod( Void.class, TestMethod.class, "shrink", float.class ) );
	
	public static final EntityType EXTENT = core.newEntityType( new IdMap(LEFT, RIGHT), new IdMap(), new IdMap(SHRINK), View.NONE );
	
	public static void shrink( Entity e, float amount )
	{
		float l = e.get(LEFT).x;
		float r = e.get(RIGHT).x;
		float center = (l + r) * 0.5f;
		float gap = (r - l) * 0.5f;
		
		e.get(LEFT).x = center - gap * amount;
		e.get(RIGHT).x = center + gap * amount;
	}
	
	@Test
	public void testMethod()
	{
		Entity e = new Entity( EXTENT );
		
		e.get(LEFT).x = 4.0f;
		e.get(RIGHT).x = 10.0f;
		
		e.execute( SHRINK, 0.5f );
		
		assertEquals( 5.5f, e.get(LEFT).x, 0.00001f );
		assertEquals( 8.5f, e.get(RIGHT).x, 0.00001f );
	}
	
}
